aicsimageio/readers/lif_reader.py

```diff
@@ -0,0 +1,751 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+import io
+import logging
+from pathlib import Path
+from typing import Dict, List, Optional, Tuple
+
+import dask.array as da
+import numpy as np
+from dask import delayed
+from lxml.etree import _Element
+from readlif import utilities
+from readlif.reader import LifFile, LifImage
+
+from .reader import Reader
+from .. import exceptions, types
+from ..buffer_reader import BufferReader
+from ..constants import Dimensions
+
+###############################################################################
+
+log = logging.getLogger(__name__)
+
+###############################################################################
+
+
+class LifReader(Reader):
+    """
+    LifReader wraps readlif.reader to provide the same reading capabilities but abstracts the specifics of using the
+    backend library to create a unified interface. This enables higher level functions to duck type the File Readers.
+
+    Parameters
+    ----------
+    data: types.FileLike
+        A string or path to the CZI file to be read.
+    chunk_by_dims: List[str]
+        The dimensions to use as the for mapping the chunks / blocks.
+        Default: [Dimensions.SpatialZ, Dimensions.SpatialY, Dimensions.SpatialX]
+        Note: SpatialY and SpatialX will always be added to the list if not present.
+    S: int
+        If the image has different dimensions on any scene from another, the dask array construction will fail.
+        In that case, use this parameter to specify a specific scene to construct a dask array for.
+        Default: 0 (select the first scene)
+    """
+
+    LIF_MAGIC_BYTE = 0x70
+    LIF_MEMORY_BYTE = 0x2a
+
+    ########################################################
+    #
+    #  Note the lif treats scenes as separate images in the lif file.
+    #  Also once a Scene/Image is loaded the image data is retrieved
+    #  2D XY plane by 2D XY plane meaning that if you have a Z stack.
+    #  concretely if you have loaded an image into img and you then
+    #  attempt to retrieve the XY plane with get_frame(c=0, t=0) and the
+    #  image has 20 z slices get_frame will assume z=0 is the slice you want.
+    #
+    ########################################################
+
+    def __init__(
+        self,
+        data: types.FileLike,
+        chunk_by_dims: List[str] = [Dimensions.SpatialZ, Dimensions.SpatialY, Dimensions.SpatialX],
+        S: int = 0,
+        **kwargs
+    ):
+        # Run super init to check filepath provided
+        super().__init__(data, **kwargs)
+
+        # Store parameters needed for _daread
+        self.chunk_by_dims = chunk_by_dims
+        self.specific_s_index = S
+        lif = LifFile(filename=self._file)
+        #  _chunk_offsets is a list of ndarrays (only way I could deal with inconsistent scene shape)
+        self._chunk_offsets, self._chunk_lengths = LifReader._compute_offsets(lif=lif)
```

Comment:
Please place the `_compute_offsets` function above the init. The _used_ function should be placed before the _using_ function.